src/portable/portable.c

```diff
@@ -1065,7 +1098,10 @@ static int attach_unit_file(
                 _cleanup_(unlink_and_freep) char *tmp = NULL;
                 _cleanup_close_ int fd = -1;
 
+                (void) mac_selinux_create_file_prepare_label(path, m->selinux_label);
+
                 fd = open_tmpfile_linkable(path, O_WRONLY|O_CLOEXEC, &tmp);
+                mac_selinux_create_file_clear(); /* Clear immediately in case of errors */
```

Comment:
one of these days we should add some magic macros, that allow a syntax like this:

```c
MAC_SELINUX_WITH_CREATE_LABEL(path, m->selinux_label) {
        fd = open_tmpfile_linkable(path, O_WRONLY|O_CLOEXEC, &tmp);
}
```

and that would make sure we reset the thing implicitly when the block is left. i.e. like we have for RUN_WITH_UMASK()

Reply:
Please don't:

I quite like the current approach by using wrapper functions that are just named with an appended `_label`.
Yes, one needs to write a separate wrapper function for each creator function (like `open_tmpfile_linkable`), but they are quite trivial to write and encapsulate two implementation details, which I do not like repeating at every caller:
1. the path used to lookup the label; this is especially important for creator functions creating a temporary file, where the label should be looked up for the final path.
2. the file type (directory, regular, symlink, ...).

By using a macro each contributor must be somewhat SELinux aware to decide, which path and file type to pass (potential for mistakes). Also macros are not type safe and might be misused with the extra introduced block. By just appending `_label` to a function name, the code author does not need any particular knowledge about SELinux (just that the new object should be created with a correct label).
Performance should not be an issue due to using LTO.